frontend/src/js/components/PageViewer/VirtualScroll.tsx

```diff
@@ -0,0 +1,82 @@
+import _ from "lodash";
+import React, { FC, ReactNode, useLayoutEffect, useRef, useState } from "react";
+import { CONTAINER_AND_MARGIN_SIZE } from "./model";
+import styles from "./VirtualScroll.module.css";
+
+type VirtualScrollProps = {
+  totalPages: number;
+  initialPage?: number;
+  renderPage: (pageNumber: number) => ReactNode;
+};
+
+export const VirtualScroll: FC<VirtualScrollProps> = ({
+  totalPages,
+  initialPage,
+  renderPage,
+}) => {
+  // Tweaked this and 2 seems to be a good amount on a regular monitor
+  // The fewer pages we preload the faster the initial paint will be
+  // Could possibly make it dynamic based on the visible of the container
+  const PRELOAD_PAGES = 2;
+
+  const pageHeight = CONTAINER_AND_MARGIN_SIZE;
```

Comment:
Shower though: could you store the maximum page height as metadata on the document along with page count? That way you know you’re guaranteed to get the next page even if for whatever reason it’s bigger than the hard coded bounding box

Reply:
Yeah I was thinking about this too, the thing that put me off the idea was if there's some weird PDF where a given page is *massive* then it might be preferable that we actually constrain it?